### PR TITLE
New trajectory follower system

### PR DESCRIPTION
**Pull request by caguero**
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

# 🎉 New feature

## Summary

This system adds a very basic trajectory follower applied to a model. The trajectory is performed by applying forces and torques to a link. It's possible to specify the trajectory in three ways:

 * Providing a set of waypoints.
 * Specifying a segment relative to the starting pose of the model.
 * Specifying a circle relative to the starting pose of the model.

Note that trajectories are in 2D.

## Test it

Run the `trajectory_follower.sdf` example to see it in action with a box.

You can see a more interesting example here.

trajectory_follower

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [X] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/systems/trajectory_follower/TrajectoryFollower.cc [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    // that all entities have been created when Configure is called
+    this->dataPtr->Load(_ecm, this->dataPtr->sdfConfig);
+    this->dataPtr->initialized = true;
+  }
+
+  // Nothing to do.
+  if (this->dataPtr->localWaypoints.empty())
+    return;
+
+  //this->dataPtr->modelPose = this->dataPtr->link.WorldPose(_ecm);
```
Comment (iche033) [inline src/systems/trajectory_follower/TrajectoryFollower.cc:275]: remove?

File: src/systems/trajectory_follower/TrajectoryFollower.cc [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  this->dataPtr->modelPose = ignition::gazebo::worldPose(
+    this->dataPtr->link.Entity(), _ecm);
+  // (*this->dataPtr->modelPose).Pos().Z() = 0;
```
Comment (iche033) [inline src/systems/trajectory_follower/TrajectoryFollower.cc:278]: remove?

File: src/systems/trajectory_follower/TrajectoryFollower.cc [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  if (_sdf->HasElement("loop_forever"))
+    this->loopForever = _sdf->Get<bool>("loop_forever");;
+
+  // Parse the optional <force> element.
+  if (_sdf->HasElement("force"))
+    this->forceToApply = _sdf->Get<double>("force");
+
+  // Parse the optional <torque> element.
+  if (_sdf->HasElement("torque"))
+    this->forceToApply = _sdf->Get<double>("torque");
```
Comment (iche033) [inline src/systems/trajectory_follower/TrajectoryFollower.cc:219]: `forceToApply` -> `torqueToApply`

Comment (iche033) [reply src/systems/trajectory_follower/TrajectoryFollower.cc:219]: after fixing this, I made both `<force>` and `<torque>` in demo world file to be `10` and the box moves nicely without too much overshooting when rotating

File: src/systems/trajectory_follower/TrajectoryFollower.hh [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  ///                      The default value is 2 deg.
+  /// <force>: The force to apply at every plugin iteration in the X direction
+  ///          of the link (N). The default value is 60.
+  /// <torque>: The torque to apply at every plugin iteration in the Yaw
+  ///           direction of the link (Nm). The default value is 50.
+  /// <line>: Element that indicates the model should travel in "line" mode.
+  ///         The block should contain the relative direction and distance from
+  ///         the initial position in which the vehicle should move, specified
+  ///         in the world frame.
+  ///           <direction>: Relative direction (degrees) in the world frame for
```
Comment (iche033) [inline src/systems/trajectory_follower/TrajectoryFollower.hh:77]: I think most of our angles are specified in radians? For the `<pose>` element in particular, there is a bool [degrees](http://sdformat.org/spec?ver=1.9&elem=model#pose_degrees) attribute to indicate if it's degrees or not. Don't have strong opinion on this though.

**Review by iche033 — COMMENTED** [review]
works well for me after fixing the force that's applied to objects. 
left a couple of other minor suggestions.

File: examples/worlds/trajectory_follower.sdf [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            <specular>1 0 0 1</specular>
+          </material>
+        </visual>
+      </link>
+
+      <plugin
+        filename="ignition-gazebo-trajectory-follower-system"
+        name="ignition::gazebo::systems::TrajectoryFollower">
+        <link_name>box_link</link_name>
+        <loop_forever>true</loop_forever>
```
[the remark this answers is not in the document]
Comment (caguero) [reply examples/worlds/trajectory_follower.sdf]: Changed in [f082490](https://github.com/ignitionrobotics/ign-gazebo/pull/1332/commits/f082490540e4e27aa58065479729b5605044bec7).

File: src/systems/trajectory_follower/TrajectoryFollower.cc (hunk 3 above)
Comment (caguero) [reply src/systems/trajectory_follower/TrajectoryFollower.cc]: Good catch, changed in [f082490](https://github.com/ignitionrobotics/ign-gazebo/pull/1332/commits/f082490540e4e27aa58065479729b5605044bec7).

File: src/systems/trajectory_follower/TrajectoryFollower.hh (hunk 4 above)
Comment (caguero) [reply src/systems/trajectory_follower/TrajectoryFollower.hh:77]: Changed to radians in [f082490](https://github.com/ignitionrobotics/ign-gazebo/pull/1332/commits/f082490540e4e27aa58065479729b5605044bec7).

Comment (iche033) [reply src/systems/trajectory_follower/TrajectoryFollower.hh]: thanks, I updated doc in [0927042](https://github.com/ignitionrobotics/ign-gazebo/pull/1332/commits/09270428724b15dab8c568aa44d87434961264b4)

**Review by iche033 — APPROVED** [review]
looks good to me.

**Comment by iche033** [comment]
there's a windows build error: 
```
gazebo\src\systems\trajectory_follower\TrajectoryFollower.cc(168,20): error C2065: 'M_PI': undeclared identifier
```

**Comment by iche033** [comment]
pushed a fix in [9a1ddd9](https://github.com/ignitionrobotics/ign-gazebo/pull/1332/commits/9a1ddd9dabfd61b2d6e1498486cbb8fcb6da78a4)

**Comment by chapulina** [comment]
I'm porting this forward in #1380 and it doesn't work out of the box in Garden. I'm looking into it.

I also noticed that the plugin was added without a test. We should add one.

**Pull request merged**